src/v/datalake/translation/translation_probe.h

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#pragma once
+
+#include "metrics/metrics.h"
+#include "model/fundamental.h"
+
+namespace datalake {
+
+class translation_probe final {
+public:
+    // Note: Do not forget to register new causes in
+    // register_invalid_record_metric.
+    enum class invalid_record_cause {
+        /// Failed to resolve the Kafka schema for the record. This covers the
+        /// cases where magic byte is missing from the record or it references a
+        /// non-existent schema in the registry.
+        failed_kafka_schema_resolution,
+        /// Failed to translate the record data according to the schema fetches
+        /// from schema registry to an equivalent Iceberg schema.
+        failed_data_translation,
+        /// Failed to ensure the table schema matches the inferred Iceberg
+        /// schema.
+        failed_iceberg_schema_resolution,
+    };
+
+public:
+    explicit translation_probe(model::ntp ntp);
+
+public:
+    void increment_invalid_record(invalid_record_cause cause) {
+        counter_ref(cause)++;
+    }
+
+    constexpr size_t& counter_ref(invalid_record_cause cause) {
```

Comment:
i'm not sure i really get the point of constexpr here

Reply:
https://github.com/redpanda-data/redpanda/pull/25116